### PR TITLE
feat: milestone-date-formatting

### LINKED ISSUE #8611: [FEATURE] Milestones date formatting
### Description

Would we be able to make the formatting on milestone dates optional, or even let users choose between displaying a year, the month and the year, or the date, month and the year. I have been looking at updating mine and for some of my milestones it doesn't really make sense to have a specific date, or even a month.

### Screenshots

image

It would be good if I could have the following here either per milestone or globally.

Some examples:

January, 2007
JAN 2007
2007
etc etc

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by sital002**
## Fixes Issue

closes #8611 
## Changes proposed

```account/manage/milestones``` date formatting.

~~Formatting options:~~
~~dd/MM/yyyy
January/dd/yyyy
Jan/dd/yyyy~~

Edit : New Format options 
* dd/mm/yyyy ```23/01/2020```
* mm/yyyy ```09/2020```
* year ```2020```

## Check List (Check all the applicable boxes) 

- [x] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

## Screenshots

image
image

### CODE REVIEW

**Review by amandamartin-dev — COMMENTED** [review]
Hello! Thank you for your contributions - I have a few questions...

In the issue for this, the OP is asking for options to not have a granular date that also includes a day.  What they have asked for is to be able to have 
day,month,year
year only
month/year only
Can you show me if the original request was changed so I may test against the issue?

Edit: disregard the second comment, this is fine - confused myself
-Secondly, I see some inconsistency in your format option declarations.  For example, you have dd/MM/yyyy for the first option but then Write January/dd/yyyy for the second.  This is confusing to mix generic indicators with full string examples.  Can you explain further the approach?-

**Comment by sital002** [comment]
> In the issue for this, the OP is asking for options to not have a granular date that also includes a day. What they have asked for is to be able to have day,month,year year only month/year only Can you show me if the original request was changed so I may test against the issue?
> 
My apologies I missed the optional part. Just to be clear the format options should be :
``` day/month/year```
``` year only ```
``` month/year only ```
And the user should be able to hide or show the milestone date.
Let me make the necesaary changes and hit you up.

**Comment by amandamartin-dev** [comment]
> My apologies I missed the optional part. Just to be clear the format options should be :
> ` day/month/year`
> `year only`
> `month/year only`
> And the user should be able to hide or show the milestone date.
> Let me make the necesaary changes and hit you up.

No worries! This is all part of the collaboration process. So in reading the issue description, it is asking for 2 things:
1. To have a choice in how the date is displayed if they choose to display it
2. To make the date of the milestone optional

If you believe the issue to be unclear, you can also go back to the OP and have a discussion with them, but this is my interpretation.  It can also be improved in the future if folks want more choices, etc. So these will likely be a good place to start.

**Comment by sital002** [comment]
I have the milestone date formatting options and now the user can choose between 
```22/01/2022```
```09/2022```
```2022```
>To make the date of the milestone optional
>
I asked it in the issue the milestone date doesn't have to be optional.

**Review by amandamartin-dev — COMMENTED** [review]
Awesome! Thanks for adjusting the dates.  Next question I have is that this update appears to affect all milestone displays via dropdown.  In the issue, the OP is asking to be able to set each milestone display individually.

If I am misinterpreting this code let me know - I am unable to run this locally at the moment.

**Comment by sital002** [comment]
> Awesome! Thanks for adjusting the dates. Next question I have is that this update appears to affect all milestone displays via dropdown. In the issue, the OP is asking to be able to set each milestone display individually.
> 
I have added the date formatting for individual milestones.

> If I am misinterpreting this code let me know - I am unable to run this locally at the moment.
>
Can you elaborate a bit more or maybe provide some screenshots of the error that you are facing?

**Comment by amandamartin-dev** [comment]
> > Awesome! Thanks for adjusting the dates. Next question I have is that this update appears to affect all milestone displays via dropdown. In the issue, the OP is asking to be able to set each milestone display individually.
> 
> I have added the date formatting for individual milestones.
> 
> > If I am misinterpreting this code let me know - I am unable to run this locally at the moment.
> 
> Can you elaborate a bit more or maybe provide some screenshots of the error that you are facing?

No error, I was working on my work laptop and I cannot run docker there.  I will take a look at this now later today. Thanks for the clarification

**Review by amandamartin-dev — CHANGES_REQUESTED** [review]
Hello! In testing this update, it does not appear to affect the display of the milestones.  I tested adding one of each type and do not see the format reflected in the dashboard or public view.  Please see screenshots

Tested in gitpod running in chrome

public profile
milestone-public

logged in dashboard view
milestone-dashboard

**Comment by sital002** [comment]
Strange It's working fine on my machine. Is there any error in the console? 

[Screencast from 26-09-23 20:29:56.webm](https://github.com/EddieHubCommunity/BioDrop/assets/93700622/02004b6f-e12e-484c-878c-6a4af38000f1)

Screenshot from 2023-09-26 20-35-02

**Comment by amandamartin-dev** [comment]
> Strange It's working fine on my machine. Is there any error in the console?
> 
>  Screencast.from.26-09-23.20.29.56.webm 
> Screenshot from 2023-09-26 20-35-02

let me test again - i switched to your branch, but perhaps it wasnt' loading the updated code

**Comment by amandamartin-dev** [comment]
@sital002 Tested again and confirmed the updated code from your branch is present.  No errors, but my dropdown choice does not save on new milestone creation or edit.  I do not have time to debug this further today, but I don't know quite why we are seeing two different scenarios. I will try to take a look again later this week if no one else is able to come throuhg and verify the updates.

**Comment by amandamartin-dev** [comment]
@sital002 tested once more today after making sure the main branch was merged in.  Still not able to see the date format choice persist after choosing.  We will have to wait for another maintainer to test to see if they notice anything I am not.  
Testing env - gitpod on chrome
no errors in console or terminal output

**Review by kumarsonsoff3 — DISMISSED** [review]
Works well on my system, tested it on the Gitpod chrome env & removed the unwanted empty spaces from the other files included. 

Here is the public profile view of the newly created test milestones. The same shows in the authenticated dashboard view.
image
Edited existing milestones. 
image

File: components/user/UserMilestone.js
Comment (eddiejaoude) [inline components/user/UserMilestone.js:10]: We have a date format function in `services/utils` already, maybe this is better in there also, so it can be reused

Example `services/utils/dateFormat.js`

File: pages/account/manage/milestone/[[...data]].js
Comment (eddiejaoude) [inline pages/account/manage/milestone/[[...data]].js:256]: I think this can be a single line?

```suggestion
                      onChange={(e) => setdateFormat(e.target.value)}
```

File: components/user/UserMilestone.js
Comment (eddiejaoude) [reply components/user/UserMilestone.js:10]: The name is too genetic, maybe something like `shortenDate` as we are making it shorter?

Comment (sital002) [reply components/user/UserMilestone.js]: I have moved the formatDate function to ```services/utils/dateFormat.js``` and renamed it to ```shortenDate```

File: components/user/UserMilestones.js
Comment (sital002) [inline components/user/UserMilestones.js:10]:
```suggestion
```

File: pages/account/manage/milestone/[[...data]].js
Comment (eddiejaoude) [reply pages/account/manage/milestone/[[...data]].js:257]: Resolved but I don't see a comment or changes?

Comment (sital002) [reply pages/account/manage/milestone/[[...data]].js]: Oops sorry I accidentally discard it now it's done.

**Review by eddiejaoude — APPROVED** [review]
Looks good, great collaboration everyone! Thank you 👍

**Pull request merged**
